mrp_production_force_bom/models/mrp_production.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2023 Quartile Limited
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, models
+from odoo.exceptions import ValidationError
+
+
+class MrpProduction(models.Model):
+    _inherit = "mrp.production"
+
+    def _check_unexpected_products(
+        self, move_raw_product_ids, bom_product_ids, error_messages
+    ):
+        unexpected_products = move_raw_product_ids - bom_product_ids
+        if unexpected_products:
+            error_messages.append(
+                _("'%s': Some components are not part of the BOM.") % self.name
+            )
+        return error_messages
+
+    def _check_missing_products(
+        self, move_raw_product_ids, bom_product_ids, error_messages
+    ):
+        missing_products = bom_product_ids - move_raw_product_ids
+        if missing_products:
+            error_messages.append(
+                _(
+                    "'%s': Some components from the BOM are missing"
+                    " in the raw materials."
+                )
+                % self.name
+            )
+        return error_messages
+
+    @api.model
+    def _force_bom(self):
+        param = (
+            self.env["ir.config_parameter"]
+            .sudo()
+            .get_param("mrp_production_force_bom.force_bom")
+        )
+        if param == "1":
+            return True
+        return False
+
+    def _skip_bom_check(self):
+        """Extend this method as necessary to control whether or not to check the
+        components against BoM.
+        """
+        self.ensure_one()
+        return False
+
+    def action_confirm(self):
+        if not self._force_bom():
+            return super(MrpProduction, self).action_confirm()
+        error_messages = []
+        for production in self:
+            if production._skip_bom_check():
+                continue
+            if not production.bom_id:
+                error_messages.append(
+                    _("'%s': You cannot proceed without setting a BOM.")
+                    % production.name
+                )
+                continue
+            bom_product_ids = set(
+                production.bom_id.mapped("bom_line_ids.product_id.id")
+            )
+            move_raw_product_ids = set(production.move_raw_ids.mapped("product_id.id"))
```

Comment:
I suppose this does not prevent users from adding an existing component product in the MO as another line, does it? How should we handle this case? Check number of lines?